packages/@aws-cdk/cfn-include/lib/cdk-include.ts

```diff
@@ -0,0 +1,221 @@
+   * as they did in the original CloudFormation template file.
+   *
+   * @default true
+   */
+  readonly preserveLogicalIds?: boolean;
+}
+
+/**
+ * A construct that allows you to include an existing CloudFormation template file into your CDK application.
+ * You can retrieve any of the resources defined in the template file by calling the {@link getResource} method.
+ * Any modifications you make on the returned resource objects will be reflected in the resulting CDK template.
+ *
+ * @see getResource
+ */
+export class CdkInclude extends core.Construct {
+  private readonly resources: { [logicalId: string]: core.CfnResource } = {};
+  private readonly cfnType2L1Class: { [cfnType: string]: string };
+  private readonly template: any;
+  private readonly preserveLogicalIds: boolean;
+
+  constructor(scope: core.Construct, id: string, props: CdkIncludeProps) {
+    super(scope, id);
+
+    // read the template into a JS object
+    this.template = fs.readJsonSync(props.templateFile);
+
+    // check for any sections of the template we don't recognize
+    for (const section of Object.keys(this.template)) {
+      switch (section) {
+        case 'AWSTemplateFormatVersion':
+        case 'Description':
+        case 'Resources':
+          // currently, we only support these 3 sections
+          break;
+        default:
+          throw new Error(`Encountered an unsupported section in the template: '${section}'`);
+      }
+    }
+
+    // read the map of CloudFormation type to their L1 class created by the build script
+    this.cfnType2L1Class = fs.readJsonSync(path.join(__dirname, '..', 'cfn-types-2-classes.json'));
+    this.preserveLogicalIds = props.preserveLogicalIds ?? true;
+
+    // instantiate all resources as CDK L1 objects
+    for (const logicalId of Object.keys(this.template.Resources || {})) {
+      this.getOrCreateResource(logicalId);
+    }
+  }
+
+  /**
+   * Returns the low-level CfnResource from the template with the given logical ID.
+   * Any modifications performed on that resource will be reflected in the resulting CDK template.
+   *
+   * The returned object will be of the proper underlying class;
+   * you can always cast it to the correct type in your code:
+   *
+   *     // assume the template contains an AWS::S3::Bucket with logical ID 'Bucket'
+   *     const cfnBucket = cfnTemplate.getResource('Bucket') as s3.CfnBucket;
+   *     // cfnBucket is of type s3.CfnBucket
+   *
+   * If the template does not contain a resource with the given logical ID,
+   * an exception will be thrown.
+   *
+   * @param logicalId the logical ID of the resource in the CloudFormation template file
+   */
+  public getResource(logicalId: string): core.CfnResource {
+    const ret = this.resources[logicalId];
+    if (!ret) {
+      throw new Error(`Resource with logical ID '${logicalId}' was not found in the template`);
+    }
+    return ret;
+  }
+
+  private getOrCreateResource(logicalId: string): core.CfnResource {
```

Comment:
Why do we need this "getOrCreate" logic? It seems this method is called only once for each resource as the template it loaded. I don't see a need for idempotency here...

Reply:
We need it, because it's actually recursive (we need to call it for expressions like `Ref`, `Fn::GetAtt`, etc. that refer to logical IDs, as the logical IDs could have changed because of `preserveLogicalIds` set to `false`).